### PR TITLE
Tilt Maze

### PR DESCRIPTION
**Pull request by imgntn**
This PR is for a Tilt Maze  -- get the ball to the end!

To test:
1. Run this script https://rawgit.com/imgntn/hifi/tiltmaze/unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js
2. Move the tilt maze (with hand controllers is best)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: cebc6937a33aef31edd59e9c8d0f9d62e262497e available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by cozza13** [comment]
that works well.

**Comment by imgntn** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: cebc6937a33aef31edd59e9c8d0f9d62e262497e available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by zzmp** [comment]
- This is awesome.
- Sometimes I can "win" by going behind the goal.
- Some of the demarcated paths are too narrow to use. There shouldn't be paths there.
- It would be nice if the bottom were translucent so that when I overrotate I still have a view of the ball.

**Comment by zzmp** [comment]
Doesn't appear to clean up it's entities when the script is stopped.

**Comment by imgntn** [comment]
@zzmp thanks for the feedback -- re: cleanup -- for item spawners i've been making cleanup an option, which is off by default.  otherwise people expect things to stay after they create them and when they quit interface, the stuff disappears... any thoughts on what the default should be?

**Comment by zzmp** [comment]
I was surprised by this one because I manually deleted the tiltmaze, and a whole bunch of helper entities stuck around. I don't mind things sticking around after scripts if they are item spawners, but I don't like entities that I can't see sticking around. Maybe trigger a cleanup of everything when I delete the tiltmaze?

**Comment by imgntn** [comment]
yeah, i can hook into the delete signal for the maze to do the other cleanup

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4866ab9eadf50870a6487ee1cca80f9cc31e0311 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by imgntn** [comment]
DNM while we update model & textures

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5e57366ba717516fc076391cc21a8400bbcbe24d available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 643365bc5a41047024871bb9281eb9a00a183c00 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

File: unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+var ball, ballSpawningAnchor, ballDetector, tiltMaze, lightAtTheEnd;
+
+var MAZE_MODEL_URL = "http://hifi-content.s3.amazonaws.com/DomainContent/Home/tiltMaze/newmaze_tex-4.fbx";
+var MAZE_COLLISION_HULL = "http://hifi-content.s3.amazonaws.com/DomainContent/Home/tiltMaze/newmaze_tex-3.obj";
+var MAZE_SCRIPT = Script.resolvePath('maze.js?' + Math.random());
+
+var SCALE = 0.5;
+
+var MAZE_DIMENSIONS = Vec3.multiply(SCALE, {
+  x: 1,
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js]: coding standard: this file's indents are wrong... should be 4 spaces not 2.

File: unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    damping: BALL_DAMPING,
+    gravity: BALL_GRAVITY,
+    density: BALL_DENSITY,
+    color: BALL_COLOR,
+    dimensions: BALL_DIMENSIONS,
+    userData: JSON.stringify({
+      grabbableKey: {
+        grabbable: false
+      }
+    })
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    position: position,
+    restitution: MAZE_RESTITUTION,
+    damping: MAZE_DAMPING,
+    angularDamping: MAZE_ANGULAR_DAMPING,
+    dynamic: true,
+    density: MAZE_DENSITY,
+    script: MAZE_SCRIPT
+  }
+
+  tiltMaze = Entities.addEntity(properties)
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+var createLightAtTheEnd = function() {
+
+  var mazeProps = Entities.getEntityProperties(tiltMaze, 'position');
+
+  var up = Quat.getUp(mazeProps.rotation);
+  var down = Vec3.multiply(-1, up);
+
+  var emitOrientation = Quat.rotationBetween(Vec3.UNIT_NEG_Z, down)
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  Entities.editEntity(tiltMaze, {
+    userData: JSON.stringify({
+      tiltMaze: {
+        firstBall: ball,
+        ballSpawner: ballSpawningAnchor,
+        detector: ballDetector,
+        lightAtTheEnd: lightAtTheEnd
+      }
+    })
+  })
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+createAll();
+
+if (CLEANUP === true) {
+  Script.scriptEnding.connect(function() {
+    Entities.deleteEntity(tiltMaze);
+    Entities.deleteEntity(ball);
+    Entities.deleteEntity(ballSpawningAnchor);
+    Entities.deleteEntity(lightAtTheEnd);
+  })
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/createTiltMaze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        x: 0,
+        y: -9.8,
+        z: 0
+    };
+
+    var BALL_DIMENSIONS = Vec3.multiply(SCALE, {
+        x: 0.05,
+        y: 0.05,
+        z: 0.05
+    })
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+    var BALL_COLOR = {
+        red: 255,
+        green: 0,
+        blue: 0
+    }
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+                density: BALL_DENSITY,
+                color: BALL_COLOR,
+                dimensions: BALL_DIMENSIONS
+
+            };
+
+            this.ball = Entities.addEntity(properties);
+        },
+        destroyBall: function() {
+            var res = Entities.findEntities(MyAvatar.position, 10);
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: coding standard: variable names must be descriptive

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+            res.forEach(function(r) {
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: coding standard: variable names must be descriptive

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        },
+        testBallDistance: function() {
+            if (this.ballLocked === true) {
+                return;
+            }
+            // print('test ball distance')
+            var userData = Entities.getEntityProperties(this.entityID, 'userData').userData;
+            var data = null;
+            try {
+                data = JSON.parse(userData)
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+            } catch (e) {
+                print('error parsing json in maze userdata')
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+            }
+            if (data === null) {
+                print('data is null in userData')
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+            // print('DATA IS:'+JSON.stringify(data))
+
+            var ballPosition;
+            if (this.ball === null) {
+                this.ball = data.tiltMaze.firstBall;
+                ballPosition = Entities.getEntityProperties(data.tiltMaze.firstBall, 'position').position;
+
+            } else {
+                ballPosition = Entities.getEntityProperties(this.ball, 'position').position
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+        },
+        testWinDistance: function() {
+            if (this.ballLocked === true) {
+                return;
+            }
+            // print('testing win distance')
+            var userData = Entities.getEntityProperties(this.entityID, 'userData').userData;
+            var data = null;
+            try {
+                data = JSON.parse(userData)
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+            } catch (e) {
+                print('error parsing json in maze userdata')
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+            }
+            if (data === null) {
+                print('data is null in userData')
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+                return;
+            }
+
+            var ballPosition;
+            if (this.ball === null) {
+                this.ball = data.tiltMaze.firstBall;
+                ballPosition = Entities.getEntityProperties(data.tiltMaze.firstBall, 'position').position;
+            } else {
+                ballPosition = Entities.getEntityProperties(this.ball, 'position').position
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+            var separation = Vec3.distance(ballPosition, ballDetectorPosition);
+            // print('separation from win:' + separation)
+            if (separation < BALL_DETECTOR_THRESHOLD) {
+                // print('BALL IS IN DETECTOR, MAKE A NEW ONE')
+                this.ballLocked = true;
+                this.destroyBall();
+                this.playVictorySound();
+                Script.setTimeout(function() {
+                    _this.ballLocked = false;
+                    _this.createBall()
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+                }, 1500)
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: missing ;

File: unpublishedScripts/DomainContent/Home/tiltMaze/maze.js [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+            } else {
+
```
Comment (ZappoMan) [inline unpublishedScripts/DomainContent/Home/tiltMaze/maze.js]: why have this blank section here?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9b6889e237a8f5438e5c9b7e98d3daba8bb54707 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f62777af9694b57f44c05acc781b0034724da78e available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 751b991522da16d27e7ba88dce3a6ffad6315f1e available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8d5834475a139c8ff695bc44f78351e9efe7c8a3 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 16bebc437b1b143a199fa830dc53cb8601e2d8e7 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 16bebc437b1b143a199fa830dc53cb8601e2d8e7 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 16bebc437b1b143a199fa830dc53cb8601e2d8e7 available at: https://gist.github.com/fdfd3523aac51d847ec0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7170.dmg

**Pull request merged**